### PR TITLE
Improve speed of dump_connections() in layer_impl.h

### LINKED ISSUE #3142: Slow DumpLayerConnections()
**Is your feature request related to a problem? Please describe.**
Execution of DumpLayerConnections() is probably too slow (see code below).

**Describe alternatives you've considered**
Given the present implementation of DumpLayerConnections(), a solution would be to profile the implementation and improve speed.

Another possible solution (in addition to the previous one) could be to add another version of DumpLayerConnections() that accepts a collection of connections as input parameter. Something similar to:

my_conn = nest.GetConnections(layer1, layer2, ...)
nest.DumpConnections(my_conn)

**Additional context**
The following code takes 1m20s to execute with NEST 3.6 (git master head from february 12th) Python 3.8.12, RockyLinux 8.5

import nest

pos = nest.spatial.grid(shape = [100,100] )

input_l = nest.Create('iaf_psc_alpha', positions=pos)
layer_0 = nest.Create('iaf_psc_alpha', positions=pos)

conn_neur = {'rule':'pairwise_bernoulli', 'mask': {'grid':{'shape':[10,10]}} }
syn_0 = {'synapse_model': 'static_synapse'}

nest.Connect(input_l, layer_0, conn_neur, syn_0)

nest.DumpLayerConnections(input_l, layer_0, 'static_synapse', 'conn.txt')

### PR DESCRIPTION
**Pull request by xavierotazuGDS**
Instead of calling get_connections() inside a loop for every source node, it calls get_connections() once for all connections between layers.

The code considers that:
1) source nodes positions (src_vec) is ordered by source node.
2) connectome is ordered by source node.

fixes #3142

### CODE REVIEW

**Comment by xavierotazuGDS** [comment]
This is a proposed solution for issue #3142

**Comment by xavierotazuGDS** [comment]
This implementation only works for non-MPI and in some cases (not all) of MPI. Hence, DO NOT USE !!!!

**Comment by heplesser** [comment]
@xavierotazuGDS Thank you for your PR! Since there still seems to be a problem with the code as of your last message, I have converted it to "Draft" status. You can change that back again later.

**Comment by xavierotazuGDS** [comment]
I have uploaded a new version. No bugs now on MPI. I do not change the Draft status because I am not sure how to do it.

**Comment by xavierotazuGDS** [comment]
Status changed from "Draft" to "Open".

**Comment by terhorstd** [comment]
Thanks for the update, xavierotazuGDS!

There are/were still some formatting issues (mostly regarding white-spaces as far as I see), but they need to be fixed to keep code in a consistently formatted state  (see [Contribute to NEST](https://nest-simulator.readthedocs.io/en/stable/developer_space/index.html)).

Since you seem to be a first time contributor, please also send a [Contributor Agreement](https://nest-simulator.readthedocs.io/en/stable/_downloads/9b65adbdacba6bfed66e68c62af4e308/NEST_Contributor_Agreement.pdf) to info@nest-initiative.org so we can accept and correctly acknowledge your contribution.

**Comment by terhorstd** [comment]
I just saw, that you already *did* send the CA! My appologies.

**Comment by xavierotazuGDS** [comment]
terhorstd,

I believe I solved all my code style issues. The other issues are in methods I have not modified (they are clang-format issues from the original code). Anyway, I could try to solve them (I will need some time).

**Comment by xavierotazuGDS** [comment]
I took a look at clang-issues in methods I have not modified, but suggested changes are so big code blocks that I would not know what I am doing. Hence, I prefer not to modify clang-format issues outside the method I modifed.

**Comment by terhorstd** [comment]
The version of the code formatter changed some time back, this is why there are complaints. Everything is already fixed in the main branch, which your branch is several commits behind (see on the main page of your repo "… behind"). Try to merge those into your branch, that should fix all spots that you didn't touch.

File: nestkernel/layer_impl.h [hunk 1]
```diff
@@ -309,48 +309,53 @@ Layer< D >::dump_connections( std::ostream& out,
...
+    out << source_node_id << ' ' << target_node_id << ' ' << weight << ' ' << delay;
+
+    Layer< D >* tgt_layer = dynamic_cast< Layer< D >* >( target_layer.get() );
+
+    out << ' ';
+    const long tnode_lid = tgt_layer->node_collection_->get_lid( target_node_id );
+    assert( tnode_lid >= 0 );
+    tgt_layer->compute_displacement( source_pos, tnode_lid ).print( out );
+    out << '\n';
 }
```
Comment (otcathatsya) [inline nestkernel/layer_impl.h:359]: Missing closing bracket here

Comment (xavierotazuGDS) [reply nestkernel/layer_impl.h:360]: corrected

**Review by heplesser — COMMENTED** [review]
@xavierotazuGDS My apologies for taking a long time to react to your PR. Your code is largely fine, but can be come quite a bit cleaner and simpler by using C++'s `find_if()` function. I have created  PR against your repo that implements it. Once you accept that PR in your repo, it will automatically pop up here and I will approve.

We would like to integrate this improvement into the release candidate for NEST 3.8 within a few days.

**Comment by heplesser** [comment]
@otcathatsya I have worked a bit more on the improved dump_connection(). You can find my current status here: https://github.com/heplesser/nest-simulator/tree/xdump. Could you run a subset of your benchmarks again on that branch to ensure I did not introduce performance regressions?

**Comment by heplesser** [comment]
> @otcathatsya I have worked a bit more on the improved dump_connection(). You can find my current status here: https://github.com/heplesser/nest-simulator/tree/xdump. Could you run a subset of your benchmarks again on that branch to ensure I did not introduce performance regressions?

My branch is now merged here, so you can work on @xavierotazuGDS's branch for this PR.

**Review by heplesser — APPROVED** [review]
@xavierotazuGDS The merge worked and this looks good now, so I approve. Before merging, we need to wait for benchmark data by @otcathatsya.

**Comment by otcathatsya** [comment]
Sorry for the hold-up, here are the benchmarks for the latest commit:
dump_conns

**Review by otcathatsya — APPROVED** [review]

**Comment by heplesser** [comment]
@otcathatsya Thank you for the data—and good they look, so here we merge ...

**Pull request merged**